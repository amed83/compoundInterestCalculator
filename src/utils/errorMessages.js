import { AMOUNT_CONSTRAINTS } from "../constants/constants";

export const errorMessages = {
  emptyInput: `This field can't be empty`,
  wrongAmount: `The amount must be between ${AMOUNT_CONSTRAINTS.lowLimit}£ and ${AMOUNT_CONSTRAINTS.upLimit}£`
};
