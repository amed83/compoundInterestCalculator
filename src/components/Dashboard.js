import React, { useEffect } from "react";
import { SLIDER_MARKS } from "../constants/constants";
import { errorMessages } from "../utils/errorMessages";
import {
  StyledInput,
  StyledContainer,
  StyledSliderContainer,
  StyledSlider,
  StyledErrorText,
  StyledTotal,
  StyledSliderTitle,
  StyledMonthlyRepay
} from "./styles";

import { calculateWithCompoundInterest } from "../utils/compoundFormula/compoundFormula";

const Dashboard = () => {
  const [months, setMonths] = React.useState(1);
  const [monthlyPayment, setMonthlyPayment] = React.useState(500);
  const [amount, setAmount] = React.useState(500);
  const [total, setTotal] = React.useState(monthlyPayment);
  const [error, setError] = React.useState(null);

  const errorMessage = message => {
    setTotal(0);
    setMonthlyPayment(0);
    return setError(`${message}`);
  };

  const validation = input => {
    setError(null);
    if (!input) {
      return errorMessage(errorMessages.emptyInput);
    }

    if (input < 500 || input > 10000) {
      return errorMessage(errorMessages.wrongAmount);
    }
    setAmount(input);
  };

  const changeLoanAmount = e => {
    const { value } = e.target;
    validation(parseInt(value));
  };

  const changeMonths = (e, val) => {
    setMonths(val);
  };

  React.useEffect(() => {
    if (!amount || !months || error) {
      return;
    }
    const total = calculateWithCompoundInterest(amount, months);
    setTotal(total);
    setMonthlyPayment(total / months);
  }, [amount, months]);

  function valuetext(monthsInput) {
    return `${monthsInput} Months`;
  }

  return (
    <StyledContainer>
      <StyledInput
        data-test="input"
        defaultValue={amount}
        id="standard-basic"
        label="Amount"
        onChange={e => changeLoanAmount(e)}
      />
      <StyledErrorText data-test={"error"}>{error}</StyledErrorText>

      <StyledSliderContainer classes={{ root: "root-class" }}>
        <StyledSliderTitle id="discrete-slider-always" gutterBottom>
          Months
        </StyledSliderTitle>

        <StyledSlider
          classes={{ root: "root-class" }}
          data-test="slider"
          aria-labelledby="discrete-slider-always"
          getAriaValueText={valuetext}
          defaultValue={1}
          onChange={(e, val) => changeMonths(e, val)}
          step={1}
          min={1}
          max={48}
          valueLabelDisplay="on"
          marks={SLIDER_MARKS}
        />
      </StyledSliderContainer>
      <StyledTotal>Total: £{total}</StyledTotal>
      <StyledMonthlyRepay>
        Monthly Payment: £{monthlyPayment ? monthlyPayment.toFixed(1) : 0}
      </StyledMonthlyRepay>
    </StyledContainer>
  );
};

export default Dashboard;
