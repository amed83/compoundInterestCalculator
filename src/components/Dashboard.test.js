import Dashboard from "./Dashboard";
import React from "react";
import { errorMessages } from "../utils/errorMessages";

describe("Dashboard ", () => {
  let wrapper;

  let props;
  const useStateMock = jest.spyOn(React, "useState");
  // const useEffectMock = jest
  //   .spyOn(React, "useEffect")
  //   .mockImplementation(f => f());

  let validationMock;
  const months = 1;
  const setMonths = jest.fn();
  const amount = 500;
  const setAmount = jest.fn();
  const setError = jest.fn();
  const error = "";
  let totalWithCompoundInterestMock;
  let changeMonthsMock;
  useStateMock.mockImplementation(initialState => [months, setMonths]);
  beforeEach(() => {
    // jest.spyOn(React, "useEffect").mockImplementation(f => f());

    validationMock = jest.fn();
    props = {
      validation: validationMock,
      mockSetState: "",
      totalWithCompoundInterest: totalWithCompoundInterestMock,
      changeMonths: changeMonthsMock
    };
    totalWithCompoundInterestMock = jest.fn();
    changeMonthsMock = jest.fn();
    wrapper = shallow(<Dashboard {...props} />);
  });

  test("<Dashboard> should render correctly ", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  test("Triggering onChange in the Slider will cause a change on months state", () => {
    const slider = wrapper.find('[data-test="slider"]');
    const event = "event";

    slider.simulate("change", event, 2);
    expect(setMonths).toHaveBeenCalledWith(2);
    useStateMock.mockImplementation(initialState => [amount, setAmount]);
  });

  test("Triggering onChange in the TextFiled cause a change on the amount state", () => {
    const inputField = wrapper.find('[data-test="input"]');

    const event = {
      target: {
        value: 500
      }
    };

    inputField.simulate("change", event);
    expect(setAmount).toHaveBeenCalledWith(event.target.value);
  });

  test("An error will displayed if the amount is less of the min limit", () => {
    useStateMock.mockImplementation(state => [error, setError]);
    wrapper = shallow(<Dashboard {...props} />);
    const inputField = wrapper.find('[data-test="input"]');
    const event = {
      target: {
        value: 400
      }
    };
    inputField.simulate("change", event);
    expect(setError).toHaveBeenCalledWith(errorMessages.wrongAmount);
  });

  test("An error will displayed if the input filed is left empty", () => {
    wrapper = shallow(<Dashboard {...props} />);
    const inputField = wrapper.find('[data-test="input"]');
    const event = {
      target: {
        value: ""
      }
    };
    inputField.simulate("change", event);
    expect(setError).toHaveBeenCalledWith(errorMessages.emptyInput);
  });
});
