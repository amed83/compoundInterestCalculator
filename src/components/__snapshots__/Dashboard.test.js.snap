// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard  <Dashboard> should render correctly  1`] = `
"<Styled(WithStyles(ForwardRef(Container)))>
  <Styled(WithStyles(ForwardRef(TextField))) data-test=\\"input\\" defaultValue={1} id=\\"standard-basic\\" label=\\"Amount\\" onChange={[Function: onChange]} />
  <styled.p data-test=\\"error\\">
    1
  </styled.p>
  <Styled(WithStyles(ForwardRef(Container))) classes={{...}}>
    <styled.h3 id=\\"discrete-slider-always\\" gutterBottom={true}>
      Months
    </styled.h3>
    <Styled(WithStyles(ForwardRef(Slider))) classes={{...}} data-test=\\"slider\\" aria-labelledby=\\"discrete-slider-always\\" getAriaValueText={[Function: valuetext]} defaultValue={1} onChange={[Function: onChange]} step={1} min={1} max={48} valueLabelDisplay=\\"on\\" marks={{...}} />
  </Styled(WithStyles(ForwardRef(Container)))>
  <styled.h2>
    Total: £
    1
  </styled.h2>
  <styled.h3>
    Monthly Payment: £
    1.0
  </styled.h3>
</Styled(WithStyles(ForwardRef(Container)))>"
`;
